art/classifiers/GPy.py

```diff
@@ -0,0 +1,254 @@
+
+    def class_gradient(self, x, label=None, logits=False, eps=0.0001):
+        """
+        Compute per-class derivatives w.r.t. `x`.
+
+        :param x: Sample input with shape as expected by the model.
+        :type x: `np.ndarray`
+        :param label: Index of a specific per-class derivative. If an integer is provided, the gradient of that class
+                      output is computed for all samples. If multiple values as provided, the first dimension should
+                      match the batch size of `x`, and each value will be used as target for its corresponding sample in
+                      `x`. If `None`, then gradients for all classes will be computed for each sample.
+        :type label: `int` or `list`
+        :param logits: `True` if the prediction should be done at the logits layer.
+        :type logits: `bool`
+        :return: Array of gradients of input features w.r.t. each class in the form
+                 `(batch_size, nb_classes, input_shape)` when computing for all classes, otherwise shape becomes
+                 `(batch_size, 1, input_shape)` when `label` parameter is specified.
+        :rtype: `np.ndarray`
+        """
+        eps = 0.00001
+        grads = np.zeros((np.shape(x)[0], 2, np.shape(x)[1]))
+        for i in range(np.shape(x)[0]):
+            # get gradient for the two classes GPC can maximally have
+            for c in range(2):
+                ind = self.predict(x[i].reshape(1, -1))[0, c]
+                sur = self.predict(np.repeat(x[i].reshape(1, -1),
+                                             np.shape(x)[1], 0) + eps*np.eye(np.shape(x)[1]))[:, c]
+                grads[i, c] = ((sur-ind)*eps).reshape(1, -1)
+        if not(label is None):
+            return grads[:, label, :].reshape(np.shape(x)[0], 1, np.shape(x)[1])
+        else:
+            return grads
+
+    def get_activations(self, x, layer, batch_size):
+        raise NotImplementedError
+
+    def loss_gradient(self, x, y):
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Sample input with shape as expected by the model.
+        :type x: `np.ndarray`
+        :param y: Correct labels, one-vs-rest encoding.
+        :type y: `np.ndarray`
+        :return: Array of gradients of the same shape as `x`.
+        :rtype: `np.ndarray`
+        """
+        eps = 0.00001
+        grads = np.zeros(np.shape(x))
+        for i in range(np.shape(x)[0]):
+            # 1.0 - to mimic loss, [0,np.argmax] to get right class
+            ind = 1.0-self.predict(x[i].reshape(1, -1))[0, np.argmax(y[i])]
+            sur = 1.0-self.predict(np.repeat(x[i].reshape(1, -1), np.shape(x)[1], 0)
+                                   + eps*np.eye(np.shape(x)[1]))[:, np.argmax(y[i])]
+            grads[i] = ((sur-ind)*eps).reshape(1, -1)
+        return grads
+
+    def predict(self, x, logits=False, batch_size=128):
```

Comment:
Please remove `batch_size` from signature and docstring.

Reply:
is removed.